keps/0020-controller-overhaul.md

```diff
@@ -0,0 +1,160 @@
+            * [Pros](#pros-1)
+            * [Cons](#cons-1)
+         * [Have a separate CRD just for maintaining commands to execute a plan](#have-a-separate-crd-just-for-maintaining-commands-to-execute-a-plan)
+            * [Pros](#pros-2)
+            * [Cons](#cons-2)
+         * [Introduce new sub-resource for triggering plan executions](#introduce-new-sub-resource-for-triggering-plan-executions)
+            * [Pros](#pros-3)
+            * [Cons](#cons-3)
+         * [User Stories](#user-stories)
+            * [As an operator (person) I want to be able to run a custom plan on my KUDO operator in order to trigger non-default plan](#as-an-operator-person-i-want-to-be-able-to-run-a-custom-plan-on-my-kudo-operator-in-order-to-trigger-non-default-plan)
+            * [As a stateful service operator (person) I want to be able to run backup on my stateful service when a backup plan is defined.](#as-a-stateful-service-operator-person-i-want-to-be-able-to-run-backup-on-my-stateful-service-when-a-backup-plan-is-defined)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Since the inception of KUDO, being able to manually run a plan was one of the core features. After KEP-18 (controller overhaul) we changed the way we store information about executed plans practically removing the ability to execute custom plan in an easy way through kubectl. Goal of this KEP is to propose a way how to re-introduce that functionality and provide a good user experience.
+
+## Motivation
+
+### Goals
+
+Operator will be able to trigger plan of given name through CLI
+Controller won’t allow multiple plans running in parallel
+
+### Non-Goals
+
+Provide a history of plan executions older that the one last run per plan.
+
+## Proposal
+
+There are multiple ways we can do this. The goal of the KEP is to choose the one with the right balance of added complexity and with a good enough UX. The main challenge with executing plans is that kubernetes API in its essence is declarative while starting a plan is an imperative action (a command) so it does not fit very nicely into the ecosystem.
+
+### Update the status of Instance with name of the plan you want to run, let controller pick it up
+
+This is by far the simplest solution but probably not the one with cleanest design. We just need an API call that will change the value of instance.Status.PlanStatus.Status to Pending for the status where instance.Status.PlanStatus.Name is equal to the plan you want to run (this is what current controller does when executing one of the built-in plans). Although it’s very easy to do this, it’s not very kubernetes idiomatic way of doing things especially because Status should never be updated from a client and it should just capture the state of the object.
```

Comment:
> Although it’s very easy to do this...

Is it though? I guess, you mean that the _cli_ will override the _Instance_, setting the new `instance.Status.PlanStatus.Status`, correct? However, the instance controller also overrides the `Instance` during reconciliation, meaning we will have a race condition here?

Reply:
I'll be more specific here. The implementation is easy meaning you need something like:

```
instance.Status = newStatus
client.Status().Update(context.TODO(), instance)
```

the client updating status hopefully should not be a problem, since there should be admission hook in place that won't let you update status when there is another plan in progress...